### PR TITLE
Draft for adding multi level select capabilities

### PR DESCRIPTION
**Pull request by KasMA1990**
This is a draft to gather feedback, and do some testing, so it's not considered ready for merging.

I would both like feedback on the UX of it (so please run the `select` example, and play around with it), and also on the approaches in the code. I _think_ I've matched the existing style pretty well, but let me know if there's something you would like to be different.

If the feedback is good, I'll start looking at tests and documentation.

I haven't looked at navigating back and forth with the arrow keys yet, but I would really like to get that in before we merge this. But I'll let that wait until the other stuff is ready I think 🙂

Implements most of #382

### CODE REVIEW

File: prompt/_generic_list.ts [hunk 1]
```diff
@@ -111,31 +111,52 @@ export abstract class GenericList<T, V, S extends GenericListSettings<T, V>>
       this.options = this.settings.options.slice();
     } else {
       this.options = this.settings.options
-        .filter((option: GenericListOptionSettings) =>
-          match(option.name) ||
-          (option.name !== option.value && match(option.value))
-        )
-        .sort((a: GenericListOptionSettings, b: GenericListOptionSettings) =>
-          distance(a.name, input) - distance(b.name, input)
-        );
```
Comment (KasMA1990) [inline prompt/_generic_list.ts:120]: All the changes in this file are just pulling code into smaller functions, so I can reuse bits of it in the `Select` module

File: prompt/select.ts [hunk 2]
```diff
@@ -14,10 +14,15 @@ import { GenericPrompt } from "./_generic_prompt.ts";
...
-export type SelectOption = GenericListOption;
+export interface SelectOption extends GenericListOption {
+  options?: SelectValueOptions;
+}
 
 /** Select option settings. */
-export type SelectOptionSettings = GenericListOptionSettings;
+export interface SelectOptionSettings extends GenericListOptionSettings {
+  options: SelectValueSettings;
+  indentLevel: number;
```
Comment (KasMA1990) [inline prompt/select.ts:24]: I messed around with different ways of rendering search results that are nested in categories, and the best solution I came up includes attaching an indentation level to the settings of each value. More details below.

File: prompt/select.ts [hunk 3]
```diff
@@ -85,12 +113,31 @@ export class Select<S extends SelectSettings = SelectSettings>
...
-    line += isSelected ? `${this.settings.listPointer} ` : "  ";
+
+    if (this.isCurrentlySearching()) {
+      line += this.settings.indent.repeat(item.indentLevel);
+      line += isSelected ? `${this.settings.listPointer} ` : "  ";
+    } else {
+      let pointer = "";
+      if (Select.itemIsBackButton(item)) {
+        // TODO: Make configurable
+        pointer = blue(`⮠`);
```
Comment (KasMA1990) [inline prompt/select.ts:124]: I feel it's a nice touch that categories and the back button get their own pointers, but I assume they should be configurable, like the normal list pointer 😄

File: prompt/select.ts [hunk 4]
```diff
@@ -99,6 +146,167 @@ export class Select<S extends SelectSettings = SelectSettings>
...
 
+  protected async submit(): Promise<void> {
+    const itemToSubmit = this.options[this.listIndex];
+
+    if (Select.itemIsBackButton(itemToSubmit)) {
+      const previousLevel = this.#parentCategories.pop();
+      if (typeof previousLevel === "object") {
+        this.options = previousLevel.options;
+        if (this.isCurrentlyInsideCategory()) {
+          this.listIndex = 2;
```
Comment (KasMA1990) [inline prompt/select.ts:157]: When you're in a category, the first two items are the back button and a separator, so the default index should skip those elements.

File: prompt/select.ts [hunk 5]
```diff
@@ -99,6 +146,167 @@ export class Select<S extends SelectSettings = SelectSettings>
...
+        this.listIndex = 2;
+      } else {
+        this.options = this.settings.options.slice();
+        this.clampListIndex();
+      }
+    } else {
+      const newOptions = this.itemsToSearchIn()
+        .filter((option) => Select.matchesItemOrSubItems(input, option));
+
+      this.options = this.buildSearchResult(input, newOptions);
```
Comment (KasMA1990) [inline prompt/select.ts:240]: The way I make search results look good, is by taking the original options that were found, and then building a bunch of new options around those. This lets categories have their own items which are disabled, while their children appear enabled below.

File: prompt/select.ts [hunk 6]
```diff
@@ -99,6 +146,167 @@ export class Select<S extends SelectSettings = SelectSettings>
...
+
+        this.options = this.itemsInsideCategory(categoryToRevertTo);
+        this.listIndex = 2;
+      } else {
+        this.options = this.settings.options.slice();
+        this.clampListIndex();
+      }
+    } else {
+      const newOptions = this.itemsToSearchIn()
+        .filter((option) => Select.matchesItemOrSubItems(input, option));
```
[the remark this answers is not in the document]
Comment (KasMA1990) [reply prompt/select.ts:238]: Of course, this code now means select options will never be sorted, which could be seen as a regression. So we could detect if there are no categories in the search results, and then sort only when that's the case.

We can also add sorting between options that are in the same category, but trying to sort all search results across different categories seems like it will just be confusing.

**Comment by c4spar** [comment]
Nice! I tried the example and it works already pretty well 👍 Tomorrow i should have time too look over your changes.

File: prompt/select.ts (hunk 3 above)
Comment (c4spar) [reply prompt/select.ts:124]: Yeah configurable would be nice 👍

File: prompt/select.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (c4spar) [reply prompt/select.ts:238]: We could try to sort all items by string distance and all category names alphabetically below the items.

Initial:
```
? Pick a color ⌕ 
group1
  matchedItem2
  matchedItem1
matchedItem2
matchedItem1
group2
  matchedItem1
  matchedItem2
```

After search:
```
? Pick a color ⌕ matchedItem
matchedItem1
matchedItem2
group1
  matchedItem1
  matchedItem2
group2
  matchedItem1
  matchedItem2
```

[the remark this answers is not in the document]
Comment (c4spar) [reply prompt/select.ts:238]: but maybe we should put the sorting in general behind an option 🤔

[the remark this answers is not in the document]
Comment (KasMA1990) [reply prompt/select.ts:238]: I think that risks putting the most relevant results the furthest away (if say `group2` contains an exact match). Maybe something like this is better:
- Divide the search results into "groups", where:
  - Any non-category item is part of the group of items that share the same parent category
  - Any non-category item that does not have a parent category becomes its own group, with no other members
- Define a group's distance to the search string as the smallest distance of all its group members to the input string, and sort the groups by this distance
- Internally in a group, also sort members by their distance to the input, relative to each other.

In other words, the example might look like this:
Initial:
```
? Pick a color ⌕ 
group1
  aaa
  abb
abb
aab
group2
  bbb
  aac
```

After search:
```
? Pick a color ⌕ aa
group1
  aaa
  abb
aab
group2
  aac
abb
```

[the remark this answers is not in the document]
Comment (KasMA1990) [reply prompt/select.ts:238]: And without sorting, the ordering becomes essentially random, which I also don't know if is desirable. So I'm not sure if it's worth allowing to turn it off 🤷

**Comment by KasMA1990** [comment]
> Nice! I tried the example and it works already pretty well 👍 Tomorrow i should have time too look over your changes.

Awesome, thanks! I'm pretty happy with how this is getting off the ground too 😄

Aside from the code changes, these are the UX changes that are worth having a look at I think:
- When you navigate into a category, the title of the category is displayed at the top
  - What style should this use?
  - If you navigate inside another category, they show as "firstCategoryName/secondCategoryName"
  - If you navigate into a third (and beyond) category, all middle category names get truncated out of the path, ala "firstCategoryName/../thirdCategoryName"
- In the list of selectable items, categories look like all other items, _except_ that they have their own pointer
  - Should they have their own pointer in the first place, and if so, what should it be?
  - Should they look distinct in some other way?
    - If you're mixing categories and regular items, it might be nice to make them distinct, but if you e.g. have the first level only as categories, and the second level only as regular items, it might look funny?
- The "Back" button also has its own pointer, and also a separator between it and the other options
  - Should it look different?
- When searching, what should categories look like?
  - Currently categories also get highlighted if they match the input text, but I don't think they should
  - They also just look white at the moment when they don't match the input, which also looks weird I think.

File: prompt/select.ts [hunk 7]
```diff
@@ -99,6 +146,167 @@ export class Select<S extends SelectSettings = SelectSettings>
...
+          italic(
+            this.nameOfParentCategory(0) + "/" + this.nameOfParentCategory(1),
+          ),
+        );
+        break;
+      
+      default: 
+        categoryText = dim(
+          italic(
+            this.nameOfParentCategory(0) + "/../" +
```
Comment (c4spar) [inline prompt/select.ts:204]: I think i would use `›` instead of `/` for the breadcrumbs:
- `Foo › .. › Bar`
- or `› Foo › .. › Bar`

File: prompt/select.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (c4spar) [reply prompt/select.ts:238]: I think we can implement the sorting for now and look how it feels. If it feels confusing we can add an option to disable group and general sorting.

File: prompt/select.ts (hunk 5 above)
Comment (c4spar) [reply prompt/select.ts:240]: I think for now this is fine. But i think i would like to add support for nested options also to the checkbox prompt. To do this we would need to move some of your logic into the generic list class and than i think we don't need to flat all the options. But i would do this in a second PR.

File: prompt/select.ts (hunk 7 above)
Comment (c4spar) [reply prompt/select.ts:204]: And i think we can show more items by default and maybe make it configurable in form of max items or max length.

File: prompt/select.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (KasMA1990) [reply prompt/select.ts:238]: Do you have a preference for which sorting to implement? 🙂

File: prompt/select.ts (hunk 5 above)
Comment (KasMA1990) [reply prompt/select.ts:240]: I'll stick with this then 😊 I guess in order to flatten it, navigation would need to not just work with the index of the selected item, but something more complex? Anyway, I'll let you see what you prefer 😄

File: prompt/select.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (c4spar) [reply prompt/select.ts:238]: I think you can just sort everithing for now, groups and entries.

File: prompt/select.ts [hunk 8]
```diff
@@ -14,10 +14,15 @@ import { GenericPrompt } from "./_generic_prompt.ts";
 export type SelectKeys = GenericListKeys;
 
 /** Select option options. */
-export type SelectOption = GenericListOption;
+export interface SelectOption extends GenericListOption {
+  options?: SelectValueOptions;
+}
```
[the remark this answers is not in the document]
Comment (c4spar) [reply prompt/select.ts:19]: I mean, the user should be able to add an option or a group of options but not a group with a value.
So the `SelectValueOptions` type could be updated like this:
```ts
export type SelectValueOptions = (string | SelectOption | SelectOptionGroup)[];
```
But I think we would already need to add support for nested options to the `GenericList` class for this 🤔

[the remark this answers is not in the document]
Comment (KasMA1990) [reply prompt/select.ts:19]: I'm having trouble making the compiler happy here at least, so I suspect you're right. But I'm having trouble grokking how all the types line up, and what exactly it is that makes `SelectOptionGroup` incompatible with the existing code.

File: prompt/select.ts (hunk 7 above)
Comment (KasMA1990) [reply prompt/select.ts]: I still don't know what the text styling should be though, to make sure it looks distinct enough from the actual prompting.

Comment (KasMA1990) [reply prompt/select.ts]: I also changed the behavior a bit, so it's the first elements that get removed, instead of the middle ones. That was easier 😁

File: prompt/select.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (KasMA1990) [reply prompt/select.ts:19]: Before I start down this path, I'd like to hear more from you about it. It seems like this could lead to a lot of scope creep, if it leads to a bunch of non-`Select` controls also supporting groups. Is that desirable? Can we implement this support in `GenericList`, but only expose it in `Select` for now? And is it even what you think should happen?

[the remark this answers is not in the document]
Comment (c4spar) [reply prompt/select.ts:19]: Sry for late reply! Than lets keep it  as it is for now. We can adapt the interfaces and the checkbox prompt in a second PR.

**Comment by KasMA1990** [comment]
Alright, I think I've addressed all concerns now 😊 The styling could probably use more feedback, and I'm waiting for more review before I start adding tests. So I'll let you have a look at it and test it out 🙂

You've talked about expanding this to other prompt types when this is done, but I'd like to end my involvement after this PR is merged, since I'm trying to cut down on things I have in flight at the moment 😅

**Comment by KasMA1990** [comment]
As for removing `value` from categories, a thought occurred to me as I was making new example data: if categories are allowed to have a `value`, then it could allow the final prompt selection to not only be a string, but potentially a list of strings, denoting the path. This could also be useful some times 🙂

**Comment by KasMA1990** [comment]
Hey, got any idea when you might have time to look this over? 😊 I don't need any specific time, but just "hopefully soon", "not soon", or "no idea" would be nice ☺️

**Comment by c4spar** [comment]
Sry for the late reply @KasMA1990, but I have been very busy lately.
My plan is to finish some refactorings for the prompt module first, then I will continue with this PR. But i'm almost done with it. I hope to get to it in the next few days.

**Comment by c4spar** [comment]
@KasMA1990 I will look into it now and resolve the merge conflicts.

File: prompt/select.ts (hunk 4 above)
Comment (c4spar) [reply prompt/select.ts]: I combined the backbutton and the breacrumb now. I think it looks more clean. What do you think?

File: prompt/select.ts (hunk 5 above)
Comment (c4spar) [reply prompt/select.ts]: I have now moved the logic to the generic list class to add support for nested options for the checkbox prompt as well.
I also no longer disable the groups while searching because otherwise I won't be able to select all the items from that group while searching, and I think it's also a nice behavior if you can navigate through the groups while searching.

File: prompt/select.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (c4spar) [reply prompt/select.ts]: I have now moved the logic to the generic list class and also splitted the option interfaces. So options have now the type `Array<string | SelectOption | SelectOptionGroup>`.

File: prompt/select.ts (hunk 7 above)
Comment (c4spar) [reply prompt/select.ts]: ince i combined the back button and the separator, i think it's fine if the separator looks the same as other categories.

**Comment by c4spar** [comment]
Hey @KasMA1990,

i did now some refactorings:
- moved the logic the the generic prompt class to add support for it to the checkbox prompt as well.
- show icons while searching
- no longer disable groups while search to be able to navigate for and back while searching
- splitted the option interfaces into options and group options
- changed the group icons (opened/closed folder)
- the group options no longer have a checked property, instead the group settings have a checked getter and there is now also a parially checked state for groups with ther own icon. i did not found a good icon for partial selection. currently i use the radio button icon `⦿` because it looks similar to some html checkboxes with partial selection but it's a circle and not a square.

I think it looks already pretty good and the search/sorting feels also good 👍 

But there are still a few thinks that needs to be fixed and cleaned up.
Let me know what you think 😉

File: prompt/select.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (c4spar) [reply prompt/select.ts]: I forgot that it is not possible to use interfaces as union type as long as they don't have overlapping properties with a different type. I think the only solution is to re-use the `value` property for the child options insted of the `options` property.

**Comment by KasMA1990** [comment]
This is very nice I think! My only gripe at this point is probably that it doesn't feel natural that I can't search for groups, I can only search by items. But I also think that's fine to leave for now, and then come back to later if needed.

Which things are missing now you think?

**Comment by c4spar** [comment]
Currently I am aware of the following things that are missing:
- [ ] groups are not searchable, as you noted.
- [ ] it should be possible to disable the group icons (I'm working on it right now)
- [ ] it is currently not possible to enter a space in the search, because the space key checks/unchecks an option (checkbox query only)
- [ ] it is currently not possible to submit when a group is selected because it enters the group (checkbox prompt only)
  - I'm not sure how to fix this. Maybe we should put the search behind a shortcut 🤔 or a "Submit/Done" option.
- [ ] Back button/parent group can't be checked and the checked/unchecked icon is missing (only checkbox prompt).
- [ ] tests are missing

**Comment by c4spar** [comment]
@KasMA1990 Sry for the very long time, but i think this MR is finally ready to merge 🙂 Thanks again for your work 👍

**Comment by KasMA1990** [comment]
And _thank you_ for your work as well! I'm sorry I haven't been able to help get it over the finish line, but I'm stoked it's here! 😃
